### PR TITLE
Fix sharing of sampling params in multiple seq groups

### PR DESCRIPTION
**Pull request by maxdebayser**
The sampling params object cannot be shared between multiple sequence groups even though the configuration is identical because some logits processors are stateful.

### CODE REVIEW

**Comment by njhill** [comment]
Thanks @maxdebayser for investigating this! I'm not sure that this is the best fix though. I think we'd want to avoid validating the same set of sampling parameters multiple times.

The sampling params are actually also already defensively copied per sequence group inside the engine [here](https://github.com/IBM/vllm/blob/main/vllm/engine/llm_engine.py#L483), but logits processors are intentionally omitted since they can be arbitrary and the cost of cloning on every request can be prohibitively high (depending on the processor implementation).

I feel like what should be passed in is actually something more like a logit processor factory. Which itself is stateless and can be included in a sampling params object that's reused. Then internally when each seq group is created it would use the factory to create the logits processor.

**Comment by maxdebayser** [comment]
Actually the JSONLogitsProcessor in the code above uses an entirely different code path. What is causing the error in our case is the CFGLogitsProcessor in vllm. It has a CFGFSM that rebuilds internal RegexFSMs based on the string it has seen so far.

**Comment by maxdebayser** [comment]
@njhill, here goes another fix idea: instead of writing a generic code that instantiates processors for each sequence just in case there might be a stateful one, we can make this state explicit with an extra sequence id argument to the logits processors. In this way the code stays as optimal as before and the author of a logits processor can decide how to optimize for his particular case.

I've overwritten my previous commit in this branch with this alternative implementation.

**Comment by njhill** [comment]
Thanks @maxdebayser I'm not sure about changing the logitsprocessor interface for this. I think there may be other reasons that we may want to change it, like passing token ids and/or working on batches. I don't think that individual logits processor impls should have to manage a dict like this.

> instead of writing a generic code that instantiates processors for each sequence just in case there might be a stateful one

Actually I was suggesting that we add logit processor factories as an additional samplingparams field, so they can be passed as alternative to existing logitsprocessors. This can be an abstract class with a method e.g. get_processor(), In the stateless case the factory can just hold and return a constant logitsprocessor. This may be better than a callable since we could later add a method return_proceessor() to allow for pooling.
